From: SeongJae Park <sj@kernel.org>
Date: Fri, 14 Feb 2025 17:32:47 -0800
Subject: [PATCH] mm/damon/core: put ops-handled filters to damos->ops_filters

damos->ops_filters has introduced to be used for operations layer
handled filters.  But DAMON kernel API callers can put any type of DAMOS
filters to damos->filters.  DAMON ABI users have no way other than still
using only damos->filters.  Update damos_add_filter(), which should be
used by API callers to install DAMOS filters, to add core layers and
operation layer handled filters to ->filters and ->ops_filters,
respectively.  The change is applied to both API callers and ABI users
since ABI users use the API internally.

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/core.c | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index 5a35ce60fa54..188399aa9429 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -281,9 +281,24 @@ struct damos_filter *damos_new_filter(enum damos_filter_type type,
 	return filter;
 }
 
+static bool damos_filter_for_ops(enum damos_filter_type type)
+{
+	switch (type) {
+	case DAMOS_FILTER_TYPE_ADDR:
+	case DAMOS_FILTER_TYPE_TARGET:
+		return false;
+	default:
+		break;
+	}
+	return true;
+}
+
 void damos_add_filter(struct damos *s, struct damos_filter *f)
 {
-	list_add_tail(&f->list, &s->filters);
+	if (damos_filter_for_ops(f->type))
+		list_add_tail(&f->list, &s->ops_filters);
+	else
+		list_add_tail(&f->list, &s->filters);
 }
 
 static void damos_del_filter(struct damos_filter *f)
-- 
2.39.5

